File: web/ts/tests/markdown.test.mjs
// The renderer configuration is tested against the very bundle that ships, so
// this covers the real rendering rather than a copy of it.
//
// It exercises the configuration alone, without DOMPurify, because sanitising
// needs a DOM this test has no way to provide. That is deliberate: the point
// of html:false is that the *parser* never emits attacker-controlled markup in
// the first place, so the escaping below is a real defence rather than
// something the sanitiser has to clean up afterwards.

import assert from "node:assert/strict";
import test from "node:test";

import { createRenderer, markdownOptions, sanitizeConfig } from "../../static/markdown-config.js";
import { vendorBundle } from "./vendor.mjs";

const { default: MarkdownIt } = await import(vendorBundle("markdown-it"));

const md = createRenderer(MarkdownIt);

test("renders CommonMark", () => {
  assert.match(md.render("*emphasis*"), /<em>emphasis<\/em>/);
  assert.match(md.render("**strong**"), /<strong>strong<\/strong>/);
  assert.match(md.render("# Heading"), /<h1>Heading<\/h1>/);
  assert.match(md.render("`code`"), /<code>code<\/code>/);
  assert.match(md.render("> quoted"), /<blockquote>/);
});

// The pinned GFM set, one test per extension.

test("renders GFM tables", () => {
  const html = md.render("| a | b |\n| - | - |\n| 1 | 2 |\n");
  assert.match(html, /<table>/);
  assert.match(html, /<th>a<\/th>/);
  assert.match(html, /<td>1<\/td>/);
});

test("renders GFM strikethrough", () => {
  assert.match(md.render("~~gone~~"), /<s>gone<\/s>/);
});

test("renders GFM task lists", () => {
  const html = md.render("- [ ] todo\n- [x] done\n");
  assert.match(html, /<input type="checkbox" disabled>/);
  assert.match(html, /<input type="checkbox" disabled checked>/);
  assert.match(html, /class="task-list-item"/);
  assert.match(html, /todo/);
  assert.doesNotMatch(html, /\[ \]/, "the marker itself is replaced, not left in the text");
});

test("renders GFM autolinks", () => {
  assert.match(
    md.render("see https://example.com/1 for details"),
    /<a href="https:\/\/example\.com\/1">/,
  );
  assert.match(md.render("see www.example.com"), /<a href="http:\/\/www\.example\.com">/);
  assert.match(md.render("mail dev@example.com"), /<a href="mailto:dev@example\.com">/);
});

// Where linkify-it is wider than GFM, pinned so that it is a decision and not a
// surprise. `fuzzyLink` is one switch: it cannot recognise a `www.` host, which
// GFM autolinks, without also recognising a bare one, which GFM does not. So a
// bare host is a link in the prose and absent from the API's `links` array. A
// bare IP address is neither.
test("a bare host is linkified and a bare IP address is not", () => {
  assert.match(md.render("see example.com now"), /<a href="http:\/\/example\.com">/);
  assert.doesNotMatch(md.render("ping 127.0.0.1 now"), /<a /);
});

test("renders CommonMark autolinks", () => {
  assert.match(md.render("<https://example.com/1>"), /<a href="https:\/\/example\.com\/1">/);
});

// The XSS gate. Raw HTML is escaped outright, which is stricter than GFM's
// disallowed-raw-HTML rule and safe in the same direction.
//
// The check is on the tags actually emitted, not on text that merely looks
// like markup: escaped text may well contain the characters "onclick=" without
// any of it being live.

// The tags the renderer is ever allowed to emit, read from the sanitiser's own
// allow-list rather than copied beside it. The two lists are one list: a tag
// the renderer emits and the sanitiser does not allow is deleted from what the
// reader sees, and a second copy here could only hide that by agreeing with the
// renderer while the sanitiser disagreed.
const emittableTags = new Set(sanitizeConfig.ALLOWED_TAGS);

/** tagsIn returns the names of the tags a rendered fragment really contains. */
function tagsIn(html) {
  return [...html.matchAll(/<\/?([a-zA-Z][a-zA-Z0-9]*)/g)].map((m) => m[1].toLowerCase());
}

// The gate read the other way. Above, nothing the renderer emits may fall
// outside the allow-list; here, nothing the pinned dialect renders may be
// missing from it. Only the first direction fails loudly — a tag the sanitiser
// does not allow is deleted, so the reader loses the markup and no test
// notices. Strikethrough was exactly that: the renderer emitted `<s>` and the
// allow-list named `del`, so `~~x~~` reached the page as plain text.
test("every tag the pinned dialect renders survives the allow-list", () => {
  const document = [
    "# h1", "## h2", "### h3", "#### h4", "##### h5", "###### h6",
    "*em* **strong** ~~struck~~ `code`",
    "a line\\\nbroken",
    "---",
    "> quoted",
    "- bullet", "1. numbered", "- [x] done", "- [ ] todo",
    "| a | b |\n| - | - |\n| 1 | 2 |",
    "```\nfenced\n```",
    "[link](https://example.com/1) ![image](https://example.com/i.png)",
  ].join("\n\n");

  const emitted = new Set(tagsIn(md.render(document)));
  assert.ok(emitted.has("s"), "the document exercises strikethrough");
  for (const tag of emitted) {
    assert.ok(emittableTags.has(tag), `<${tag}> is rendered but the sanitiser strips it`);
  }
});

test("emits no tag outside the allowed set", () => {
  const attacks = [
    "<script>alert(1)</script>",
    "<img src=x onerror=alert(1)>",
    "<iframe src='https://evil.example.com'></iframe>",
    '<a href="https://ok.example.com" onclick="alert(1)">text</a>',
    "<svg><use href='#x' /></svg>",
    "<style>body{display:none}</style>",
    "<object data='x'></object>",
    "<form action='https://evil.example.com'><input name=a></form>",
    "<math><mi//xlink:href=\"data:x,<script>alert(1)</script>\">",
    "<!-- <img src=x onerror=alert(1)> -->",
  ];

  for (const attack of attacks) {
    const html = md.render(attack);
    for (const tag of tagsIn(html)) {
      assert.ok(emittableTags.has(tag), `${attack} emitted a <${tag}>`);
    }
    assert.match(html, /&lt;/, `${attack} was not escaped`);
  }
});

test("emits no event-handler attribute", () => {
  // Attributes are read from the tags themselves, so escaped text saying
  // "onclick=" does not count against this.
  const html = md.render('<img src=x onerror=alert(1)><a onclick="alert(1)">x</a>');
  for (const tag of html.matchAll(/<[a-zA-Z][^>]*>/g)) {
    assert.doesNotMatch(tag[0], /\son\w+\s*=/i, `live event handler in ${tag[0]}`);
  }
});

test("does not linkify a javascript: URL into an anchor", () => {
  const html = md.render("[click](javascript:alert(1))");
  assert.doesNotMatch(html, /href="javascript:/i);
});

// A hand-written anchor is escaped rather than honoured, so its href, its text
// and any attributes it carried are all inert. This is the same rule that
// makes the API's derived links array ignore raw HTML.
//
// One divergence is worth stating rather than hiding: because the tag is
// escaped to *text*, linkify then sees the URL inside it as a bare URL and
// links it. goldmark, which keeps the tag as a raw-HTML node, does not, so a
// description written this way renders a link the links array does not list.
test("a hand-written anchor is escaped, not honoured", () => {
  const html = md.render('<a href="https://example.com/1" onclick="alert(1)">raw</a>');

  assert.match(html, /&lt;a href=/, "the tag itself is escaped");
  assert.doesNotMatch(html, /<a[^>]*onclick/i, "its attributes are not honoured");
  assert.match(html, /raw/, "its text survives as text");
  assert.equal(markdownOptions.html, false);
});

test("the pinned options are what the renderer is built with", () => {
  assert.equal(markdownOptions.html, false, "raw HTML is escaped");
  assert.equal(markdownOptions.linkify, true, "the autolink extension is on");
  assert.equal(markdownOptions.breaks, false, "GFM does not turn a single newline into a break");
  assert.equal(markdownOptions.typographer, false, "nothing beyond the pinned set");
});

test("the sanitiser allow-lists no scripting surface", () => {
  for (const tag of ["script", "iframe", "object", "embed", "style", "form", "base", "link"]) {
    assert.ok(!sanitizeConfig.ALLOWED_TAGS.includes(tag), `${tag} must not be allowed`);
  }
  for (const attr of sanitizeConfig.ALLOWED_ATTR) {
    assert.doesNotMatch(attr, /^on/i, `${attr} looks like an event handler`);
  }
  assert.equal(sanitizeConfig.ALLOW_DATA_ATTR, false);

  // The tags a description legitimately renders to are all present.
  for (const tag of ["p", "a", "code", "pre", "table", "ul", "li", "input"]) {
    assert.ok(sanitizeConfig.ALLOWED_TAGS.includes(tag), `${tag} must be allowed`);
  }
});

test("rendering is deterministic", () => {
  const source = "# Title\n\n- [x] done\n- [ ] todo\n\nSee https://example.com/1 and **this**.\n";
  const first = md.render(source);
  for (let i = 0; i < 5; i++) {
    assert.equal(md.render(source), first);
  }
});

// Two denial-of-service advisories, both remediated by the bundled markdown-it
// release and each reached through a different part of linkify. A description
// or comment is prose one workspace member writes and every other member's
// browser renders, so either is a stored denial of service rather than
// something a caller only does to itself.
//
// Both inputs are deliberately larger than the 64 KiB a single description or
// comment may hold: one issue view renders the description and every comment on
// it, so what a page feeds the renderer is the sum, not the cap. The bound is
// wall-clock and the same for both, and generous — each takes on the order of a
// hundred milliseconds here and tens of seconds on a vulnerable bundle, so the
// two are nowhere near close enough for a loaded CI runner to confuse them.
const redosBudgetMs = 2000;

/** renderMs renders source and returns how long that took, in milliseconds. */
function renderMs(source) {
  const started = process.hrtime.bigint();
  md.render(source);
  return Number(process.hrtime.bigint() - started) / 1e6;
}

// GHSA-38c4-r59v-3vqw / CVE-2026-2327, markdown-it >=13.0.0 <14.1.1: the
// linkify rule trimmed trailing asterisks off a matched link with /\*+$/, which
// backtracks catastrophically on a long run of them ending in anything else.
//
// The asterisks have to follow something linkify actually matched — a bare run
// of them is never handed to that regular expression at all, and costs nothing.
test("a long asterisk run after a link renders in linear time", () => {
  const elapsedMs = renderMs("https://example.com/" + "*".repeat(256 * 1024) + "x");
  assert.ok(elapsedMs < redosBudgetMs, `rendering a 256 KiB asterisk run took ${elapsedMs.toFixed(0)} ms`);
});

// GHSA-v245-v573-v5vm / CVE-2026-59887, linkify-it <=5.0.1, which markdown-it
// 14 depended on: the `mailto:` validator copied and rescanned the rest of the
// text at every occurrence, so a run of them costs time quadratic in its length.
test("a run of mailto prefixes renders in linear time", () => {
  const elapsedMs = renderMs("mailto::".repeat(32 * 1024)); // 256 KiB
  assert.ok(elapsedMs < redosBudgetMs, `rendering 256 KiB of mailto prefixes took ${elapsedMs.toFixed(0)} ms`);
});

File: internal/domain/markdown_test.go
package domain_test

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/tofutools/awb/internal/domain"
)

// accepted is Markdown the gate lets through unchanged.
var accepted = []string{
	"",
	"plain prose",
	"# Heading\n\n- a\n- b\n\n> quote\n\n    indented code\n",
	"**bold** _italic_ ~~struck~~ `code` and a\n\n| a | b |\n|---|---|\n| 1 | 2 |\n",
	"- [ ] todo\n- [x] done\n",

	// Every allowed link scheme, written every way a link can be written.
	"[a](https://example.com/x)",
	"[a](http://example.com/x)",
	"[a](mailto:someone@example.com)",
	"<https://example.com/x>",
	"<mailto:someone@example.com>",
	"someone@example.com",       // GFM autolink, which becomes mailto:
	"visit www.example.com now", // GFM autolink, which becomes http://
	"[a][ref]\n\n[ref]: https://example.com/x",
	"[HTTPS://EXAMPLE.COM/X](HTTPS://EXAMPLE.COM/X)", // the scheme is case-insensitive

	// No scheme at all: relative, fragment, protocol-relative, a colon after
	// something that is not a scheme, and nothing.
	"[a](./notes.md) [b](#section) [c](//example.com/x) [d](./notes:1.md) [e]()",
	"![a](./shot.png)",

	// Images: http, https and the four raster data types, base64 or not.
	"![a](https://example.com/x.png)",
	"![a](data:image/png;base64,iVBORw0KGgo=)",
	"![a](data:image/jpeg;base64,AAAA)",
	"![a](data:image/gif;base64,AAAA)",
	"![a](data:image/webp;base64,AAAA)",
	"![a](DATA:IMAGE/PNG;BASE64,AAAA)",
	"![a](data:image/png,%89PNG)",

	// A disallowed scheme is only a destination rule, never a text rule, and a
	// < that begins no tag is prose.
	"the javascript: scheme, and `<script>` in a code span",
	"```html\n<script>alert(1)</script>\n```\n",
	"    <script>alert(1)</script>\n", // an indented code block
	"a \\<span> escaped, and 3 < 5 > 2 unescaped, and <3",
	"C:\\Users\\x, git@github.com:tofutools/awb.git and key: value",
	"the file:///etc/passwd path and ssh://host/path, neither of them autolinked",
}

// rejectedLinks are destinations no link may carry.
var rejectedLinks = []string{
	"javascript:alert(1)",
	"JaVaScRiPt:alert(1)",
	"vbscript:msgbox(1)",
	"data:text/html;base64,PHNjcmlwdD4=",
	"data:image/png;base64,AAAA", // allowed on an image, never on a link
	"file:///etc/passwd",
	"ftp://example.com/x",
	"notes:1.md", // scheme-shaped, so it is a scheme, and not one of the three
	"tel:+15551234",
	"about:blank",
}

// rejectedImages are destinations no image may carry.
var rejectedImages = []string{
	"javascript:alert(1)",
	"mailto:someone@example.com", // allowed on a link, never on an image
	"data:image/svg+xml;base64,PHN2Zz4=",
	"data:text/html,<script>alert(1)</script>",
	"data:image/png", // no comma, so no media type
	"data:,AAAA",     // the default media type is text/plain
	"data:base64,AAAA",
	"data:image/png;charset=utf-8;base64,AAAA",
}

func TestValidateMarkdownAccepts(t *testing.T) {
	for _, s := range accepted {
		assert.NoError(t, domain.ValidateMarkdown("description", s), "%q", s)
	}
}

// A bare URL the GFM autolink extension picks up is a link, so it meets the
// link rule like any other. That is deliberate rather than incidental: what it
// renders as is a destination the reader can click.
func TestValidateMarkdownAppliesTheLinkRuleToAutolinkedProse(t *testing.T) {
	assertUsage(t, domain.ValidateMarkdown("description", "see ftp://example.com/x for the archive"))

	// Wrapping it takes it out of the rule, because then it is no longer a link.
	require.NoError(t, domain.ValidateMarkdown("description", "see `ftp://example.com/x` for the archive"))
}

func TestValidateMarkdownRejectsRawHTML(t *testing.T) {
	// Raw HTML is refused rather than escaped, which is what keeps script,
	// style, SVG and MathML out: every one of them is raw HTML to the parser.
	for _, s := range []string{
		"<script>alert(1)</script>",
		"<style>body{display:none}</style>",
		"<svg onload=alert(1)></svg>",
		"<math><mtext></mtext></math>",
		"<div>block</div>",
		"<p>\n\n**still a block**\n\n</p>",
		"inline <b>bold</b> html",
		"inline <img src=x onerror=alert(1)>",
		"an <a href=\"https://example.com\">anchor</a> written by hand",
		"<!-- a comment -->",
		"prose with a bare <placeholder> in it",
		"| a |\n|---|\n| <b>x</b> |\n",
		"- <span>in a list</span>\n",
	} {
		err := domain.ValidateMarkdown("description", s)
		assertUsage(t, err, "%q", s)
		assert.Contains(t, err.Error(), "raw HTML", "%q", s)
	}
}

func TestValidateMarkdownRejectsLinkSchemes(t *testing.T) {
	// Every syntax that produces a link destination goes through the same gate.
	forms := []func(string) string{
		func(d string) string { return "[a](" + d + ")" },
		func(d string) string { return "[a](<" + d + ">)" },
		func(d string) string { return "[a][ref]\n\n[ref]: " + d },
		func(d string) string { return "<" + d + ">" },
		func(d string) string { return "[![alt](https://example.com/i.png)](" + d + ")" },
	}
	for _, destination := range rejectedLinks {
		for _, form := range forms {
			s := form(destination)
			err := domain.ValidateMarkdown("description", s)
			assertUsage(t, err, "%q", s)
			assert.Contains(t, err.Error(), "link", "%q", s)
		}
	}
}

func TestValidateMarkdownRejectsImageSchemes(t *testing.T) {
	for _, destination := range rejectedImages {
		for _, s := range []string{
			"![a](" + destination + ")",
			"![a][ref]\n\n[ref]: " + destination,
			"[link](https://example.com/x) ![a](" + destination + ")",
		} {
			err := domain.ValidateMarkdown("description", s)
			assertUsage(t, err, "%q", s)
			assert.Contains(t, err.Error(), "image", "%q", s)
		}
	}
}

func TestValidateMarkdownSeesThroughMarkdownEscaping(t *testing.T) {
	// CommonMark resolves backslash escapes and character references inside a
	// destination, so the scheme is read after resolving them; a URL parser
	// ignores whitespace and control characters within a destination, so those
	// go too. Every one of these is javascript: by the time a renderer has it.
	for _, s := range []string{
		"[a](&#106;avascript:alert&#40;1&#41;)",
		"[a](&#x6a;avascript:alert(1))",
		"[a](javascript\\:alert(1))",
		"[a](<java\tscript:alert(1)>)",
		"[a](  javascript:alert(1)  )",
		"[a](java&NewLine;script:alert(1))",
	} {
		assertUsage(t, domain.ValidateMarkdown("description", s), "%q", s)
	}

	// Percent-encoding is not undone: it survives into the URL rather than
	// being removed on the way there, so %6a is a literal path character and
	// the destination is a relative reference to every browser.
	require.NoError(t, domain.ValidateMarkdown("description", "[a](%6aavascript:alert(1))"))

	// Nor is a backslash the escaping left in place: \s is a literal backslash
	// and an s, so the colon that follows separates nothing.
	require.NoError(t, domain.ValidateMarkdown("description", "[a](java\\script:alert(1))"))
}

func TestValidateMarkdownNamesTheField(t *testing.T) {
	err := domain.ValidateMarkdown("close reason", "<b>no</b>")
	assertUsage(t, err)
	assert.True(t, strings.HasPrefix(err.Error(), "close reason "), err.Error())

	// The refusal names the markup, which is the only way to find a stray < in
	// 64 KiB of prose.
	err = domain.ValidateMarkdown("description", strings.Repeat("prose. ", 200)+"a bare <placeholder> here")
	assertUsage(t, err)
	assert.Contains(t, err.Error(), `"<placeholder>"`, err.Error())

	// A long data URL is abbreviated rather than echoed whole.
	err = domain.ValidateMarkdown("description", "![a](data:text/html;base64,"+strings.Repeat("A", 500)+")")
	assertUsage(t, err)
	assert.Less(t, len(err.Error()), 200, err.Error())
}

// The gate is part of every prose field's validation, not something a caller
// opts into.
func TestProseFieldsApplyTheMarkdownGate(t *testing.T) {
	_, err := domain.ValidateDescription("<script>alert(1)</script>")
	assertUsage(t, err)

	_, err = domain.ValidateComment("[a](javascript:alert(1))")
	assertUsage(t, err)
	assert.True(t, strings.HasPrefix(err.Error(), "comment "), err.Error())

	// A close reason becomes the body of a Markdown comment, so it passes too.
	_, err = domain.ValidateCloseReason("done, see [why](javascript:alert(1))")
	assertUsage(t, err)

	got, err := domain.ValidateCloseReason("  done, see [why](https://example.com/x)  ")
	require.NoError(t, err)
	assert.Equal(t, "done, see [why](https://example.com/x)", got)
}
